File: app/src/main/java/nz/co/codebros/quakesnz/list/QuakeListViewModel.java
package nz.co.codebros.quakesnz.list;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

import java.util.List;

import io.reactivex.functions.Consumer;
import nz.co.codebros.quakesnz.core.model.Feature;
import nz.co.codebros.quakesnz.core.model.FeatureCollection;
import nz.co.codebros.quakesnz.repository.FeatureCollectionRepository;

/**
 * Created by Leandro on 28/10/2017.
 */

class QuakeListViewModel extends ViewModel {

    private final FeatureCollectionRepository repository;
    private MutableLiveData<List<Feature>> features;

    QuakeListViewModel(FeatureCollectionRepository repository) {
        this.repository = repository;
    }

    LiveData<List<Feature>> getFeatures() {
        if (features == null) {
            features = new MutableLiveData<>();
            repository.subscribe(new Consumer<FeatureCollection>() {
                @Override
                public void accept(FeatureCollection featureCollection) throws Exception {
                    features.setValue(featureCollection.getFeatures());
                }
            });
        }
        return features;
    }

    static class Factory extends ViewModelProvider.NewInstanceFactory {
        private final FeatureCollectionRepository repository;

        Factory(FeatureCollectionRepository repository) {
            this.repository = repository;
        }

        @NonNull
        @Override
        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
            return (T) new QuakeListViewModel(repository);
        }
    }
}

File: app/src/main/java/nz/co/codebros/quakesnz/list/QuakeListViewModel.kt
package nz.co.codebros.quakesnz.list

import android.content.SharedPreferences
import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.asLiveData
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import io.reactivex.disposables.CompositeDisposable
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import nz.co.codebros.quakesnz.core.data.Feature
import nz.co.codebros.quakesnz.core.extension.reactIn
import nz.co.codebros.quakesnz.core.extension.reduceIn
import nz.co.codebros.quakesnz.list.model.QuakeListEvent
import nz.co.codebros.quakesnz.list.model.QuakeListReactor
import nz.co.codebros.quakesnz.list.model.QuakeListState
import nz.co.codebros.quakesnz.list.model.reduce
import nz.co.codebros.quakesnz.map.model.QuakeMapEvent
import nz.co.codebros.quakesnz.map.model.QuakeMapModel
import nz.co.codebros.quakesnz.util.changesFlow
import javax.inject.Inject

@ExperimentalCoroutinesApi
@HiltViewModel
class QuakeListViewModel @Inject constructor(
    private val events: MutableSharedFlow<QuakeListEvent>,
    state: MutableStateFlow<QuakeListState>,
    reactor: QuakeListReactor,
    sharedPreferences: SharedPreferences,
    quakeMapModel: QuakeMapModel
) : ViewModel() {

    val liveEvents
        get() = events.asLiveData()

    val liveState = state.asLiveData()

    private val disposables = CompositeDisposable()

    init {

        events.reactIn(viewModelScope, reactor::react)
        state.reduceIn(viewModelScope, events, ::reduce)

        viewModelScope.launch { events.emit(QuakeListEvent.RefreshQuakes) }

        viewModelScope.launch {
            sharedPreferences.changesFlow()
                .filter { it == "pref_intensity" }
                .collect { events.emit(QuakeListEvent.RefreshQuakes) }
        }

        viewModelScope.launch {
            state.map { it.selectedFeature?.geometry?.coordinates }.collect {
                it?.let {
                    quakeMapModel.publish(QuakeMapEvent.OnNewCoordinates(it))
                } ?: Log.w(QuakeListViewModel::class.simpleName, "Coordinates are null.")
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        disposables.dispose()
    }

    fun refreshQuakes() {
        viewModelScope.launch { events.emit(QuakeListEvent.RefreshQuakes) }
    }

    fun selectQuake(feature: Feature) {
        viewModelScope.launch { events.emit(QuakeListEvent.SelectQuake(feature)) }
    }
}
